rust: linux bridge

cathay4t opened the pull request:
Supporting linux bridge and also its vlan filtering.

Integration test enabled. The not enabled test cases are caused by missing
support of bond, copy_mac_from, state::ignore.

There are some cargo clippy warnings, will fix them in another patch to
minimize this one.

almusil commented on rust/src/lib/ifaces/inter_ifaces.rs:
```diff
@@ -176,6 +179,47 @@ impl Interfaces {
+                    iface.iface_type(),
+                    port_name.to_string(),
+                ));
+            }
```
The for loop is repeated twice with almost identical body. It might be a good idea to extract it into separate function.

cathay4t replied:
Changed to use:

```rust
        for iface in
            self.kernel_ifaces.values().chain(self.user_ifaces.values())
        {

```

almusil commented on rust/src/lib/ifaces/inter_ifaces.rs:
```diff
@@ -176,6 +179,47 @@ impl Interfaces {
+            } else if let Some(iface) = self.kernel_ifaces.get_mut(&ctrl_name) {
+                iface.remove_port(&port_name)
+            }
+        }
```
Looking here it might not be necessary to have those two for loops and then this loop down here. You could remove the ports directly without intermediate `Vec`.

cathay4t replied:
Not sure what you are referring. The reason I use `pending_actions` is because we should not change item during iterating even that is safe to human being.

almusil replied:
Ack

almusil commented on rust/src/lib/ifaces/inter_ifaces_controller.rs:
```diff
@@ -234,3 +270,25 @@ pub(crate) fn set_ifaces_up_priority(ifaces: &mut Interfaces) -> bool {
+            if let Some(port_iface) =
+                cur_ifaces.get_iface(port_name, InterfaceType::Unknown)
+            {
+                if port_iface.iface_type() == InterfaceType::Unknown {
```
Is there really a need for this  additional check?

cathay4t replied:
Yes. The previous `get_iface()` will search kernel interfaces with the same name, the returned Interface might have valid interface type.

almusil replied:
Ack

almusil commented on rust/src/lib/ifaces/inter_ifaces_controller.rs:
```diff
@@ -234,3 +270,25 @@ pub(crate) fn set_ifaces_up_priority(ifaces: &mut Interfaces) -> bool {
+            }
+        }
+    }
+    ret
```
This whole function could be simplified using `flat_map`, but leaving it up to you.

cathay4t replied:
Unless big benefit, I intend to avoid these magics.
The [cargo clippy suggestion][1] remind me why Perl been dropped by community(too many magic tricks)


[1]: https://rust-lang.github.io/rust-clippy/master/index.html#option_map_unit_fn

almusil commented on rust/src/lib/nispor/linux_bridge.rs:
```diff
@@ -1,50 +1,153 @@
 }
+
+fn devide_by_user_hz(v: u32) -> u32 {
+    let user_hz = unsafe { libc::sysconf(libc::_SC_CLK_TCK) } as u32;
```
Please leave a safety note for future reference.

cathay4t replied:
Commented.

almusil commented on rust/src/libnm_dbus/connection/bridge.rs:
```diff
@@ -58,15 +294,82 @@ impl NmSettingBridge {
+    _other: HashMap<String, zvariant::OwnedValue>,
+}
+
+impl TryFrom<HashMap<String, zvariant::OwnedValue>>
```
Please use the `DbusDictionary` type alias.

almusil commented on rust/src/libnm_dbus/connection/bridge.rs:
```diff
@@ -58,15 +294,82 @@ impl NmSettingBridge {
+{
+    type Error = NmError;
+    fn try_from(
+        mut v: HashMap<String, zvariant::OwnedValue>,
```
Please use the `DbusDictionary` type alias.

cathay4t committed:
rust: add support of linux bridge

Supporting linux bridge and also its vlan filtering.

Integration test enabled. The not enabled test cases are caused by
missing support of bond, copy_mac_from, state::ignore.

There are some cargo clippy warnings, will fix them in another patch to
minimize this one.

Signed-off-by: Gris Ge <fge@redhat.com>
